File: C++/AlgorithmSolutionStorage/AlgorithmSeminar/21941.cpp
#include <iostream>
#include <algorithm>
#include <string>
#include <map>
using namespace std;

string origin;
int m;
int cache[1001][1001];
map<string, int> r;

//int func(string origin) {
//	int len = origin.length();
//
//	if (!len)return 0;
//	if (cache.find(origin) != cache.end())return cache[origin];
//
//	auto& ref = cache[origin];
//	ref = 0;
//
//	for (int i = 0; i < origin.length(); i++) {
//		ref = max(ref, func(origin.substr(0, i) + origin.substr(i + 1, len))+1);
//	}
//
//	for (auto recode : r) {
//		string key = recode.first;
//		int value = recode.second;
//
//		int index = origin.find(key);
//		if (index == -1)continue;
//
//		ref = max(ref, func(origin.substr(0, index) + origin.substr(index + key.length()))+value);
//	}
//	
//	return ref;
//}

int dp(int s, int f) {
	if (s >= f)return 0;

	int& ref = cache[s][f];
	if (ref != -1)return ref;

	ref = 0;
	for (int i = s; i < f; i++) {
		ref = max(ref, dp(s, i) + dp(i + 1, f) + 1);
	}

	string sub = origin.substr(s, (f-s));
	for (auto x : r) {
		string key = x.first;
		int value = x.second;

		int idx = sub.find(key);
		if (idx == string::npos)continue;
		idx += s;
		ref = max(ref, dp(s, idx) + dp(idx + key.length(), f) + value);
	}
	return ref;
}

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);
	freopen("input.txt", "r", stdin);

	cin >> origin >> m;
	for (int i = 0; i < m; i++) {
		string k; int v;
		cin >> k >> v;
		r[k] = v;
	}

	fill(&cache[0][0], &cache[origin.length() + 1][origin.length() + 1], -1);

	cout << dp(0, origin.length());
	return 0;
}

File: python/21941.py
import sys

sys.setrecursionlimit(10**6)
sys.stdin=open('python/input.txt','r')
input=sys.stdin.readline

def dp(s, f):
  global origin, r, cache

  if s>=f: return 0
  if cache[s][f]!=-1: return cache[s][f]
  sub=origin[s:f]

  cache[s][f]=0
  for i in range(s, f):
    cache[s][f]=max(cache[s][f], dp(s, i) + dp(i + 1, f) + 1)

  for recode in r:
    key=recode[0]
    value=recode[1]

    index=sub.find(key)
    if index==-1: continue
    index+=s
    cache[s][f]=max(cache[s][f], dp(s, index) + dp(index+len(key), f) + value)

  return cache[s][f]

def func(index):
  global origin

  if index>=len(origin): return 0
  if cache[index]!=-1: return cache[index]

  # cache[index]=0
  cache[index]=func(index+1)+1

  for recode in r:
    key=recode[0]
    value=recode[1]

    # index=origin[index:].find(key)
    # if index==-1: continue
    if origin[index:index+len(key)]==key:
      cache[index]=max(cache[index], func(index+len(key))+value)

  return cache[index]

origin=input()[:-1]
m=int(input())
# cache=[[-1 for _ in range(len(origin)+1)] for _ in range(len(origin)+1)]
cache=[-1 for _ in range(len(origin)+1)]
r=[]

for _ in range(m):
  r.append(list(map(str,input().split())))
  r[-1][1]=int(r[-1][1])

# print(dp(0,len(origin)))
print(func(0))
